File: Day19/numOfMinutes.js
/**
 * @param {number} n
 * @param {number} headID
 * @param {number[]} manager
 * @param {number[]} informTime
 * @return {number}
 */
var numOfMinutes = function (n, headID, manager, informTime) {

    let graph = {};

    for (let i = 0; i < manager.length; i++) {
        if (i == headID) continue;
        if (!(manager[i] in graph)) {
            graph[manager[i]] = [];

        }
        graph[manager[i]].push(i)

    }


    let q = [[headID, informTime[headID]]];
    let seen = new Set();
    let res = 0;

    while (q.length) {

        let data = q.shift();
        let emp = data[0];
        let time = data[1];
        if (!seen.has(emp)) {
            if (graph[emp]) {

                for (let e of graph[emp]) {
                    seen.add(emp)
                    res = Math.max(res, time + informTime[e]);
                    q.push([e, informTime[e] + time])
                }
            }
        }
    }

    return res
};

File: Day19/numOfMinutes.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    int numOfMinutes(int n, int headID, vector<int> &manager, vector<int> &informTime)
    {
        unordered_map<int, vector<int>> m;
        for (int i = 0; i < manager.size(); i++)
        {
            if (i == headID)
                continue;
            m[manager[i]].push_back(i);
        }

        //create quieue
        queue<pair<int, int>> q;
        q.push(make_pair(headID, 0));
        int res = 0;

        while (!q.empty())
        {
            int size = q.size();
            for (int i = 0; i < size; i++)
            {
                int emp = q.front().first;
                int time = q.front().second;

                q.pop(); //pop data

                int duration = informTime[emp] + time;
                res = max(res, duration);

                for (auto it : m[emp])
                {
                    q.push(make_pair(it, duration));
                }
            }
        }

        return res;
    }
};